### DISCUSSION TITLE
v2.0 Release Notes 🎉

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#1945] [Announcements]
![v2-release](https://github.com/skeletonlabs/skeleton/assets/1509726/dcbaf736-7a1f-47c1-ab64-90713d9de567)

Today we're happy to introduce Skeleton v2! This represents a huge step forward for Skeleton, introducing a number of critical updates, new features, and of course a handful of breaking changes. We'll cover the details for this release below.

# 🚨 Migration Guide 🚨

We've created a detailed migration guide to help you move from v1.x or v2 RC to the v2 stable release.

- [View the v2 migration guide](https://github.com/skeletonlabs/skeleton/discussions/1947)
- [View the v2 documentation site](https://www.skeleton.dev/)
- [View the v2 Changelog](https://github.com/skeletonlabs/skeleton/blob/master/packages/skeleton/CHANGELOG.md)
- [View the archived v1 documentation site](https://v1.skeleton.dev/)

# 🆕 What's New?

## New Tailwind Plugin (#1660)

![Screenshot 2023-08-10 at 4 57 08 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/8b5bd3d0-0117-4c72-947b-6e9abd418650)

Skeleton co-maintainer @AdrianGonz97 has provided one of the largest and most impactful updates - a brand new Skeleton Tailwind plugin, which is now a [standalone package](https://www.npmjs.com/package/@skeletonlabs/tw-plugin) separate from the core library. Under the hood, the plugin composes and injects all styles required, such Skeleton's Tailwind Elements, and injects those directly via Tailwind. This means you're no longer required to import and manage additional stylesheets such as `theme.css` or `skeleton.css`. Additionally, the plugin now supports the recently introduced `.ts` or `.js` formats for your Tailwind Config. The new plugin also changes the way you define and customize your themes settings, which we'll cover in more detail below.

## New Theme Updates (#1660)

Learn more about the new theme options from the updated [Themes](https://www.skeleton.dev/docs/themes) section of our documentation.

### Register and Switch Themes

![Screenshot 2023-08-10 at 5 24 35 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/231e2725-f857-46f1-b5af-41a108ab0218)

The new Skeleton v2 plugin now allows you to register one or more **preset** or **custom** themes directly within your Tailwind config. You may then define which theme you wish to be actively displayed via a `data-theme` attribute on the body element in your application. This means dynamic theme switching is not only possible, but quite trivial to implement!

### Preset Theme Enhancements

![Screenshot 2023-08-10 at 5 26 45 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/a9138e2c-52f5-4ff4-ab70-c6e1218cb530)

Preset themes have a simpler interface for enabling optional features, such as background gradients. This is provided via the `enhancements` setting in your plugin settings. Set the value to TRUE to opt into these extra features per each theme.

### New Theme Format

![Screenshot 2023-08-10 at 4 58 21 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/fc180ce0-d99e-407f-ba48-73f5aa0d2927)

Additionally, themes have been migrated to use CSS-in-JS, Tailwind's preferred format for injecting new styles via plugin. Themes still utilize CSS custom properties (aka CSS variables) like before, this merely changes how the theme data is delivered to the plugin. Typescript fans will also enjoy the new `CustomThemeConfig` type for simple type safety.

### Using the New Format

Whether you're using a preset theme or aiming to migrate a custom theme, we recommend following our full migration guide linked at the top of this page. Note that we've also update the [Theme Generator](https://www.skeleton.dev/docs/generator) to output themes in the new format.

## Wintry Theme (#1901)

![image](https://github.com/skeletonlabs/skeleton/assets/1509726/25dd7d66-52bf-491f-ab52-445121f609ee)

Along with the new theme updates, we've opted to introduce the new "Wintry" theme. An elegant blue-focused theme. Register and implement this alongside your favorite Skeleton preset themes.

## Dynamic Transitions for Components (#1533)

![Screenshot 2023-08-10 at 4 58 51 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/7a44b813-4a8b-4fcc-9f6b-86ed06a31d43)

Contributor @Mahmoud-zino has devoted a tremendous amount of time to introduce a new dynamic transition system. This allows you to modify or disable transitions for nearly every component in the library! See the new [Essentials > Transitions](https://www.skeleton.dev/docs/transitions) to learn more about how this works. This covers which components are supported, how to configure each transition, how to disable transitions, and of course how to adhere to best practices for accessibility. And yes, this dynamic transition system supports custom Svelte-based CSS or JS Svelte transitions!

## New Table of Contents Utility (#1832)

![Screenshot 2023-08-10 at 4 59 12 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/bd520a7f-578d-4dbb-9a81-bbe793e7d6c2)

Skeleton introduced a rudimentary Table of Contents component for our v1 launch, however, this never quite reached our standards of quality. Given this, we've raised this feature to the ground and began anew. With this, a new version of the feature can now be found under [Utilities > Table of Contents](https://www.skeleton.dev/utilities/table-of-contents).

The new version features a "crawler" action to scan a region of the page for headings, optionally append unique IDs to each heading, passing a reference for each heading to the component. While the updated component provides the visual interface and list of links on the page. Furthermore, we've reworked the scroll mechanisms to allow for native browser URL hashes for deep linking (ex: `/some/path#my-deeplinked-section`). This means the feature is now supported with or without the use of the Skeleton App Shell. You'll note the default styles are much less opinionated, but you'll find a wider range of customization is possible through the new style props.

## Tree View Updates (#1779)

![Screenshot 2023-08-10 at 5 00 40 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/0bb21405-dd9f-4f1f-b83d-fbeb526be2d2)

Contributor @Mahmoud-zino has been hard at work to introduce a number of user-requested features to the new [Tree View](https://www.skeleton.dev/components/tree-views). This includes a slew of selection methods, utility features, and even a means to generate tree views in a data-driven and recursive manner. Given the scope of the changes we're going to keep the `beta` label for a little while longer, but very much welcome your feedback on these changes. You can reference pending updates here:

- https://github.com/skeletonlabs/skeleton/issues/1943
- https://github.com/skeletonlabs/skeleton/issues/1944

## PurgeCSS Plugin Support

While not directly part of the library proper, Skeleton co-maintainer @AdrianGonz97 has launched his initial version of [vite-plugin-tailwind-purgecss](https://github.com/AdrianGonz97/vite-plugin-tailwind-purgecss). If you're not aware, Tailwind can provide a sub-par experience when it comes to purging unused CSS provided by third party Tailwind libraries such as Skeleton. To remedy this, the plugin helps ensure only the styles used in your production application make it to final CSS bundle. While optional, we highly encourage this for all Skeleton projects and have enabled it by default for all Skeleton CLI generated projects going forward. We've also [provide a simple guide to help you get started](https://www.skeleton.dev/docs/purgecss).

# ⚠️ Breaking Changes:

## Drawers, Modals, and Toasts Stores (#1831)

![Screenshot 2023-08-10 at 5 01 04 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/d352e983-044a-43a2-96ea-01414bede53a)

![Screenshot 2023-08-10 at 5 01 28 PM](https://github.com/skeletonlabs/skeleton/assets/1509726/36df4c9f-4071-4744-9376-ee864d831914)

We've refactored the stores for all singleton-based overlay features to help prevent known [SvelteKit SSR issues](https://github.com/skeletonlabs/skeleton/wiki/SvelteKit-SSR-Warning). The new requirements are covered in our migration guide, and of course you're welcome to reference the updated documentation page for each feature:

- Drawers: https://www.skeleton.dev/utilities/drawers
- Modals: https://www.skeleton.dev/utilities/modals
- Toasts: https://www.skeleton.dev/utilities/toasts

## Skeleton Data Table Removed (#1822)

We've now removed the long-deprecated and unlisted Skeleton Data Table feature. Going forward, Skeleton's official recommendation is to use [Svelte Simple Datatables](https://github.com/vincjo/datatables) paired with [Skeleton's Table Element](https://www.skeleton.dev/elements/tables) styles. For any holdouts wishing to keep the legacy data table feature, you can find a [zip backup here](https://github.com/skeletonlabs/skeleton/issues/1821#issuecomment-1665963040).

## Additional Breaking Changes

- The deprecated `whitelist | blacklist` props have been removed from the Autocomplete component.
#1822
- Removed all remnants of Skeleton's legacy typography system, which is no longer available in v2.
#1822
- Renamed the Paginator `offset` to `page` for better semantics.
#1822

# 🐞 Bugfixes

- Resolved a bug when Autocomplete `allow` and `deny` lists are empty.
#1825
- Resolved a style issue for `.bg-hover-primary-token` color in dark mode.
#1824
- Prevent Paginator from resetting `page` value on length change.
#1837
- InputChips now update binded value only once.
#1768

# Updates since the v2 Release Candidate

If you installed an early version of the Skeleton v2 release candidate, please be aware we've implemented a number of updates based on your feedback, including:

- The brand new "Wintry" preset theme is now available:
https://github.com/skeletonlabs/skeleton/pull/1902
- Resolved a bug in `vite-plugin-tailwind-purgecss` breaking styles in production:
https://github.com/AdrianGonz97/vite-plugin-tailwind-purgecss/releases
- The new CSS-in-JS theme format now support `properties_dark` for specifying dark mode overrides to CSS custom properties
https://github.com/skeletonlabs/skeleton/pull/1896
- Theme `properties` and `properties_dark` can now be extended with aribtrary CSS custom properties:
https://github.com/skeletonlabs/skeleton/pull/1937
- Scrollbar corners now utilize theme styles:
https://github.com/skeletonlabs/skeleton/pull/1908
- Improvements to the `beta` Tree View component:
https://github.com/skeletonlabs/skeleton/pull/1910
- The new Table of Components utility now now includes optional `prefix` and `suffix` parameters:
https://github.com/skeletonlabs/skeleton/pull/1956
- Added theme styles for the `indeterminate` checkbox style:
https://github.com/skeletonlabs/skeleton/pull/1917
- File Button now allows you to remove the `.btn` class style for a completely unstyled display:
https://github.com/skeletonlabs/skeleton/pull/1919
- Improved the documentation around the new store features for Drawers, Modals, and Stores:
https://github.com/skeletonlabs/skeleton/pull/1938
- We've rolled back a change to the Input Chips that introduced a reactivity regression:
https://github.com/skeletonlabs/skeleton/pull/1957
- Added a `regionLabel` prop to the RadioGroup component
https://github.com/skeletonlabs/skeleton/pull/1836
- A number of issues have been resolved in the Skeleton CLI (aka `create-skeleton-app` or `CSA`)
- We've made various corrections and updates to the Skeleton documentation
- The documentation website features a slightly updated App Bar with `version` navigation

# What's Next?

You may reference our current proposals for the [v3 roadmap](https://github.com/skeletonlabs/skeleton/discussions/1367). This provides a sneak peek into what we have planned for the future. The overall goal will be to start breaking up core features, such as Popups and Code Blocks to their own standalone packages. This will allow better implementation of the dependencies these rely on. Likewise we have some major updates planned for themes and the theme generator. Expect to see updates on this in the near future.

**Comment by buhodev** [comment]
Nice! I was importing the `skeleton-minimal.css` file before in my `+layout.svelte` file (`import '@skeletonlabs/skeleton/styles/skeleton-minimal.css';`) instead of the full one because I wanted my own custom scrollbar. Now I can only define the `preset: [{ name: "skeleton", enhancements: true }],`, and no minimal styles are provided, just the full ones. How can I successfully migrate? :)

**Reply by AdrianGonz97** [reply]
Hey @buhodev, you should be able to achieve the `skeleton-minimal.css` equivalent by disabling the base styles in the plugin!

```ts
// tailwind.config.js

skeleton({
	base: false,  // <-- add this
	themes: {
		preset: [
			{ name: 'skeleton', enhancements: true },
		]
	}
})
```

**Reply by buhodev** [reply]
Thanks! It worked 🎉

**Reply by buhodev** [reply]
Just in case, if I have a custom theme, do I still need to define the skeleton preset? I think that it's not necessary, because everything works as expected if I remove the preset and leave the custom theme

**Reply by AdrianGonz97** [reply]
You do not. The presets are not required and you're free to only use your custom theme.
